docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+directly to pods. So far, artifact sync requires a specification of sync
+patterns like
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
+In addition, the syncing needs to handle special cases for globbing and often
+requires a long list of sync patterns (#1807).
+
+Furthermore, builders should be able to inform Skaffold about additional sync paths (#1704).
+This will result in much faster deploy cycles, without users having to bother about these optimizations.
+
+## Design
+
+### Goals
+The sync mechanism in skaffold should be able to infer the destination location of local files.
+
+For example: given the following Dockerfile
+
+```dockerfile
+FROM nginx
+WORKDIR /var
+COPY static *.html www/
+ADD nginx.conf .
+ADD nginx.conf /etc/
+```
+
+The sync locations should be inferred as
+- `static/page1.html` -> `/var/www/page1.html` (! no static folder at destination !)
+- `static/assets/style.css` -> `/var/www/assets/style.css`
+- `index.html` -> `/var/www/index.html`
+- `nginx.conf` -> `/var/nginx.conf`, `/etc/nginx.conf`
+
+A fundamental change compared to the current sync mechanism is that one source path may by sync'ed to multiple destinations in the future.
+
+### Assumptions
+
+Builders know where to put static files in the container.
+This information needs to be pulled from the builders and made available in Skaffold.
+Currently, we know how to do this for Docker artifacts (@r2d4 suggested this in #1166).
+Whether jib and bazel builders can provide that information too, is unclear at the moment.
+
+### Interface changes
+
+There are at least two relevant changes
+
+1. When builders are queried for their dependencies, they must provide a list of destinations together with local source paths. The signature of `Builder.GetDependencies` (pkg/skaffold/build/build.go) therefore needs to change to:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       // OLD
+       // DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) ([]string, error)
+       // NEW
+       DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) (map[string][]string, error)
+   }
+   ```
+   
+   So far, `DependenciesForArtifact` returned a list of local paths (relative to the working directory of the artifact) that are input files for the container.
+   Now, `DependenciesForArtifact` serves two purposes. As before, the map keys are relative local paths of input files for the container.
+   In addition, the map values list all absolute remote destination paths where the input files are placed in the container.
+   This has to be a list, because a single input file may be copied to several locations in the container.
```

Comment:
I think it would be better to separate the inferred sync-map from `GetDependencies()`.  `GetDependencies` is called on any file change so it should be as lightweight as possible.  For some builders there are input files that are compiled/transformed/merged that won't map to a destination so we'd likely have a lot of entries that just map to an empty array.

Reply:
Yes, after thinking about this, I'm also much in favor of @briandealwis suggestion. Besides, avoiding the double responsibility for `GetDependencies` will probably save some headache in the future.
cc @balopat